### ISSUE TITLE
Realm Needs Parameters to Some Methods to Prevent Waiting

### ISSUE THREAD

**Issue by lightsighter** [StanfordLegion/legion#465]
Some methods like getting the volume of an index space can wait on Realm events internally. For some client applications like Legion this isn't acceptable, so there need to be parameters to tell Realm that it is not safe to do such things.

Edit: (@streichler) Note that this impacts some usage of the Realm API as cases where Realm would have waited internally now become fatal errors if the caller hasn't waited (either explicitly or by setting up a continuation to be performed once the event has triggered) beforehand.  The two cases are:
1. Direct queries of potentially-sparse index spaces must ensure the index space is locally valid:
```
  IndexSpace<...> is = ...;
+ is.make_valid().wait();
  bool inside = is.contains(p);  // or any other query
```
Note that use of an `IndexSpace` for asynchronous operations (e.g. copies or dependent partitioning operations) does NOT require this change - Realm will defer the operation as needed.

2. Creation of an accessor for an instance requires the instances metadata be available to the processor on which the task is running.  There are two ways to achieve this:
    a. "Pulling" the metadata in the task itself:
```
  RegionInstance inst = ...;
  Processor p = Processor::get_executing_processor(); // or use `proc` argument to Realm task if available
+ inst.fetch_metadata(proc).wait();
  AffineAccessor<...> acc(inst, ...);
```

b. Prefetching it as a precondition to task execution:
```
  RegionInstance inst = ...;
  Processor target = ...;
- target.spawn(..., precondition);
+ Event prefetch_done = inst.fetch_metadata(target);
+ target.spawn(..., Event::merge_events(precondition, prefetch_done));
```

**Comment by streichler** [comment]
@lightsighter would Legion ever make calls that tell Realm it _is_ safe to wait internally?  I'm inclined to just rip all those wait() calls out and instead insist that the caller handling the waiting when and how it needs to.

**Comment by lightsighter** [comment]
In the Legion case it's not safe to do that with user-level threading on because we take care of saving and restoring the few thread-local variables around all wait calls. So the answer in this case is no, we would never tell Realm that it is safe to wait internally.

**Comment by streichler** [comment]
Changes to both Realm and Legion merged in with commit e1dde6e.
